capsules/src/console.rs

```diff
@@ -132,46 +131,52 @@ impl<'a, U: UART> Console<'a, U> {
             app.write_buffer
                 .take()
                 .map_or(Err(ReturnCode::ERESERVE), |slice| {
-                    self.send(app_id, app, slice);
-                    Ok(true)
+                    match self.send(app_id, app, slice) {
+                        ReturnCode::SUCCESS => Ok(true),
+                        err => Err(err),
+                    }
                 })
         } else {
             Ok(false)
         }
     }
 
     /// Internal helper function for sending data for an existing transaction.
-    /// Cannot fail. If can't send now, it will schedule for sending later.
-    fn send(&self, app_id: AppId, app: &mut App, slice: AppSlice<Shared, u8>) {
+    /// If can't send now, it will schedule for sending later.
+    fn send(&self, app_id: AppId, app: &mut App, slice: AppSlice<Shared, u8>) -> ReturnCode {
         if self.tx_in_progress.get().is_none() {
             self.tx_in_progress.set(Some(app_id));
-            self.tx_buffer.take().map(|buffer| {
-                let mut transaction_len = app.write_remaining;
-                for (i, c) in slice.as_ref()[slice.len() - app.write_remaining..slice.len()]
-                    .iter()
-                    .enumerate()
-                {
-                    if buffer.len() <= i {
-                        break;
+            self.tx_buffer.take().map_or_else(
+                || panic!("Console internal consistency error"),
```

Comment:
Crash in a capsule seems bad.

Reply:
Yes.., but before this was an error that was silently dropped, now it's noted. There are 28 other panic's in capsules as well, most similar consistency issues. I think that the console capsule could be re-architected a bit to make this case impossible, but that's a different PR.

Reply:
Doesn't adding return codes give it a way to not be silently dropped but without crashing the entire system?